include/dlaf/communication/kernels/internal/all_reduce.h

```diff
@@ -0,0 +1,152 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2024, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+#include <complex>
+#include <utility>
+
+#include <mpi.h>
+
+#include <pika/execution.hpp>
+
+#include <dlaf/common/callable_object.h>
+#include <dlaf/common/eti.h>
+#include <dlaf/communication/communicator.h>
+#include <dlaf/communication/communicator_pipeline.h>
+#include <dlaf/communication/kernels/all_reduce.h>
+#include <dlaf/communication/message.h>
+#include <dlaf/communication/rdma.h>
+#include <dlaf/matrix/tile.h>
+#include <dlaf/sender/transform_mpi.h>
+#include <dlaf/sender/with_temporary_tile.h>
+
+namespace dlaf::comm::internal {
+template <class T, Device D_in, Device D_out>
```

Comment:
```suggestion
template <class T, Device DIn, Device DOut>
```
?

Reply:
They are variable.
In the beginning we were using `Device device` and `Backend backend` in templates, but at some point we switched to `Device D` and `Backend B`. TBH I don't remember the reason for the change.

Reply:
Ah, my understanding has been "they are template parameters and thus CamelCase" overriding "they are variable names and thus snake_case". I may have contributed to the current situation. And at least when I looked now in `coding.md` I actually don't see any explicit mention of it so I may just have misunderstood.

If we're changing them to follow the "variable names" convention, then it should be `d_in`?

I don't mind either way.